ibm_mq/datadog_checks/ibm_mq/queue_metric_collector.py

```diff
@@ -0,0 +1,180 @@
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+
+from six import iteritems
+
+from datadog_checks.base import AgentCheck, ensure_bytes, ensure_unicode
+from datadog_checks.ibm_mq.metrics import GAUGE
+
+from . import metrics
+
+try:
+    import pymqi
+except ImportError as e:
+    pymqiException = e
+    pymqi = None
+else:
+    # Since pymqi is not be available/installed on win/macOS when running e2e,
+    # we load the following constants only pymqi import succeed
+    SUPPORTED_QUEUE_TYPES = [pymqi.CMQC.MQQT_LOCAL, pymqi.CMQC.MQQT_MODEL]
+
+
+class QueueMetricCollector:
+    QUEUE_SERVICE_CHECK = 'ibm_mq.queue'
+    QUEUE_MANAGER_SERVICE_CHECK = 'ibm_mq.queue_manager'
+
+    def __init__(self, config, service_check, warning, send_metric, log):
+        self.config = config
+        self.service_check = service_check
+        self.warning = warning
+        self.send_metric = send_metric
+        self.log = log
+
+    def collect_queue_metrics(self, queue_manager):
+        self.discover_queues(queue_manager)
+        self.queue_manager_stats(queue_manager, self.config.tags)
+
+        for queue_name in self.config.queues:
+            queue_tags = self.config.tags + ["queue:{}".format(queue_name)]
+
+            for regex, q_tags in self.config.queue_tag_re:
+                if regex.match(queue_name):
+                    queue_tags.extend(q_tags)
+
+            try:
+                self.queue_stats(queue_manager, queue_name, queue_tags)
+                # some system queues don't have PCF metrics
+                # so we don't collect those metrics from those queues
+                if queue_name not in self.config.DISALLOWED_QUEUES:
+                    self.get_pcf_queue_status_metrics(queue_manager, queue_name, queue_tags)
+                    self.get_pcf_queue_reset_metrics(queue_manager, queue_name, queue_tags)
+                self.service_check(self.QUEUE_SERVICE_CHECK, AgentCheck.OK, queue_tags)
+            except Exception as e:
+                self.warning('Cannot connect to queue %s: %s', queue_name, e)
+                self.service_check(self.QUEUE_SERVICE_CHECK, AgentCheck.CRITICAL, queue_tags)
+
+    def discover_queues(self, queue_manager):
+        queues = []
+        if self.config.auto_discover_queues:
+            queues.extend(self._discover_queues(queue_manager, '*'))
+
+        if self.config.queue_patterns:
+            for pattern in self.config.queue_patterns:
+                queues.extend(self._discover_queues(queue_manager, pattern))
+
+        if self.config.queue_regex:
+            if not queues:
+                queues = self._discover_queues(queue_manager, '*')
+            keep_queues = []
+            for queue_pattern in self.config.queue_regex:
+                for queue in queues:
+                    if queue_pattern.match(queue):
+                        keep_queues.append(queue)
+            queues = keep_queues
+
+        self.config.add_queues(queues)
```

Comment:
Seems not right to modify `self.config` from `QueueMetricCollector`. 

Maybe we can store the `queues` directly on `QueueMetricCollector` instance ?

Reply:
it was not even right before splitting the file!

Reply:
Yes, indeed :)